extensions/amp-ad/0.1/amp-ad-ui.js

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use baseInstance file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+
+/**
+ * Ad display state.
+ * @enum {number}
+ */
+export const AdDisplayState = {
+  /**
+   * The ad has not been laid out, or the ad has already be unlaid out
+   */
+  NOT_LAID_OUT: 0,
+
+  /**
+   * The ad has been laid out, but runtime haven't received any response from
+   * the ad server.
+   */
+  LOADING: 1,
+
+  /**
+   * The ad has been laid out, and runtime has received render-start msg from
+   * ad server.
+   * Not used now.
+   */
+  LOADED_RENDER_START: 2,
+
+  /**
+   * The ad has been laid out, and runtime has received no-content msg from
+   * ad server.
+   */
+  LOADED_NO_CONTENT: 3,
+};
+
+export class AmpAdUIHandler {
+
+  /**
+   * @param {!AMP.BaseElement} baseInstance
+   */
+  constructor(baseInstance) {
+    /** @private {!AMP.BaseElement} */
+    this.baseInstance_ = baseInstance;
+
+    /** {number} */
+    this.state = AdDisplayState.NOT_LAID_OUT;;
+  }
+
+  /**
+   * TODO(@zhouyx): Add ad tag to the ad.
+   */
+  init() {
+  }
+
+  /**
+   * Exposed function to ad that enable them to set UI to correct display state
+   * @param {number} state
+   */
+  setDisplayState(state) {
+    if (this.state == AdDisplayState.NOT_LAID_OUT) {
+      // Once unlayout UI applied, only another layout will change the UI again
+      if (state != AdDisplayState.LOADING) {
+        return;
+      }
+    }
+    switch (state) {
+      case AdDisplayState.LOADING:
+        this.displayLoadingUI_();
+        break;
+      case AdDisplayState.LOADED_RENDER_START:
+        this.displayRenderStartUI_();
+        break;
+      case AdDisplayState.LOADED_NO_CONTENT:
+        this.displayNoContentUI_();
+        break;
+      case AdDisplayState.NOT_LAID_OUT:
+        this.displayUnlayoutUI_();
+        break;
+      default:
+        dev().assert(state, 'state is not supported');
```

Comment:
How would this ever fail? You should be doing `dev().error('state is not supported')`, I think.